### PR TITLE
SEQNG-241 Pass events as an output instead of relying on Queue.enqueueOne

### PR DESCRIPTION
**Pull request by jluhrs**
This PR is to have an initial review. I still need to fix the tests.

### CODE REVIEW

**Comment by cquiroz** [comment]
I've tried to follow the changes but I'd probably need much more understanding on the details of the implementation to comment. So I'd rather pass on judging this PR, perhaps @jdnavarro has a much better grasp on this

As a general comment I'm worried about ignoring tests. The engine has too few tests already for my tests and ignoring more is probably not a good idea. I understand that your are trying to make all the tests run, right?

**Comment by jdnavarro** [comment]
> As a general comment I'm worried about ignoring tests. The engine has too few tests already for my tests and ignoring more is probably not a good idea. I understand that your are trying to make all the tests run, right?

I'm also waiting for the tests to be updated in order to help me understand how it will work now.

**Comment by jluhrs** [comment]
The PR is now in a good state. Tests were fixed.

File: modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/Engine.scala [hunk 1]
```diff
@@ -19,10 +19,11 @@ object Engine {
 
   def empty[A]: Engine[A] = Engine(Map.empty)
 
-  implicit val engineFunctor = new Functor[Engine] {
-    def map[A, B](q: Engine[A])(f: A => B): Engine[B] =
-      Engine(q.sequences.mapValues(_.map(f)))
-  }
+// This fails to compile with the error "not found: type $anon"
```
Comment (cquiroz) [inline modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/Engine.scala]: So we don't need to rewrite this?

File: modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/package.scala [hunk 2]
```diff
@@ -273,80 +302,73 @@ package object engine {
...
   /** Traverse a process with a stateful computation. */
-  private def mapEvalState[F[_]: Monad: Catchable, A, S, B](
-    fs: Process[F, A], s: S, f: A => StateT[F, S, B]
-  ): Process[F, B] = {
-    def go(fs: Process[F, A], s: S): Process[F, B] =
-      Process.eval(fs.unconsOption).flatMap {
-        case None         => Process.halt
-        case Some((h, t)) => Process.eval(f(h).run(s)).flatMap {
-          case (s, a) => Process.emit(a) ++ go(t, s)
+  private def mapEvalState[A, S, B](
```
Comment (cquiroz) [inline modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/package.scala]: Very nice. As an extended challenge you could abstract the types for `Option` and `Task` 😄

**Comment by cquiroz** [comment]
after an explanation live of the code I think I now understand it much better, nice job

File: modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/package.scala (hunk 2 above)
Comment (jluhrs) [inline modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/package.scala]: I thought of not using Option, but pass an empty Process when no one was needed. Maybe I should? The code could be a bit cleaner.

**Pull request merged**